### PR TITLE
Banners + Banner Checkpoints = Framework Prioritized Banners 

### LINKED ISSUE #10: Create a Notification Banner Manager Concept
# Feature Request

## Description
Create and provide the banner manager such that a banner order is managed when banners are rendered by the children content. This should be provided to the model and slide-panel disclosed content and should be available at a component-level.

The naming should be scoped down to distinguish the difference between application notifications and banner (terra-alert) concepts.

### Proposed Implementation
Rendered Banners will register to the nearest `BannerCheckpoint`.
```js
const parentExample = () => (
    <>
        <h1>Title</h1>
         <BannerCheckpoint>
             <Example>
        </BannerCheckpoint>
    </>
);
```
Each checkpoint will provide a `BannerRegistrationContext` to provide API to a register and unregister alert props.

```js
const bannerRegistration = {
     register: (uniqueId, alertProps) => { // add banner to nearest checkpoint },
     unregister: (uniqueId) => { // remove banner from checkpoint },
};
```
Then, provide consumes with a `Banner` component to render when a banner, which is then registered with the Checkpoint, and will be displayed above the children content for component, slide-panel, and modal content.

```js
import Banner from 'terra-application/lib/banner'; 

const Example = () => (
    <>
        <Banner type="alert" />
        <div>example content</div>
    </>
);
```

This would render similarly to
```
|----------------------------------------|
|   Title                                |
|----------------------------------------|
|   Alert Banner                         |
|----------------------------------------|
|  example content                       |
|                                        |
|                                        |
|----------------------------------------|
```

### Banner Checkpoint Props
- `priorityOrder` the priority order to render the banners in. Needs include:
    - [OSC order](https://app.zeplin.io/project/59b1bfeae515657fa8268300/screen/5ace2c83a326b0511ea2d6cd):  ALERT, ERROR, WARNING, ADVISORY, INFO
    - Fusion: ALERT, ERROR, WARNING, GAP_CHECKING, OUTSIDE_RECORDS, INFO, SUCCESS,
- `children` 

### Banner Props
Props should match or be a condensed version of the props available for `terra-alert`.

### PR DESCRIPTION
**Pull request by emilyrohrbough**
### Summary

Create and provide the banner manager such that a banner order is managed when banners are rendered by the children content.

Closes #10 

### Deployment Link

https://terra-applic-banner-ale-gmkbu1.herokuapp.com/

### Testing

- Unit tests
- Wdio tests

Thank you for contributing to Terra.
@cerner/terra

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -97,8 +97,10 @@
     "wdio": "tt-wdio --config node_modules/terra-toolkit/config/wdio/wdio.conf.js --locales=['en','de']"
   },
   "dependencies": {
+    "@testing-library/react": "^10.0.4",
```
Comment (tbiethman) [inline package.json]: Should move to devDependencies I reckon?

File: src/banner/Banner.jsx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    if (bannerRegistration && bannerRegistration.registerBanner) {
+      bannerRegistration.registerBanner(uuid, props);
+    }
+
+    return () => {
+      if (bannerRegistration && bannerRegistration.unregisterBanner) {
+        bannerRegistration.unregisterBanner(uuid, props.type);
+      }
+    };
+  }, [bannerRegistration, props]);
```
Comment (tbiethman) [inline src/banner/Banner.jsx]: I _think_ the props object will always be different if this component is updated due to a parent render. You might want to spread the props out. Can you verify?

File: src/banner/BannerCheckpoint.jsx [hunk 3]
```diff
@@ -0,0 +1,125 @@
+import React from 'react';
```
Comment (mjhenkes) [inline src/banner/BannerCheckpoint.jsx]: what do you think about renaming this `BannerProvider`

File: src/application-base/ApplicationBase.jsx [hunk 4]
```diff
@@ -134,9 +135,11 @@ const ApplicationBase = ({
                     }}
                   >
                     <ApplicationLoadingOverlayProvider>
-                      <Suspense fallback={<ApplicationLoadingOverlay isOpen />}>
-                        {children}
-                      </Suspense>
+                      <BannerCheckpoint fitToParentIsDisabled={fitToParentIsDisabled}>
```
Comment (mjhenkes) [inline src/application-base/ApplicationBase.jsx]: We need to discuss where we want these banners in the application hierarchy. I don't think it hurts to have them here as a backup but primarily i'd imagine these show up in page layout content and in modals.

File: src/banner/Banner.jsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+const Banner = (props) => {
+  const bannerRegistration = React.useContext(BannerRegistrationContext);
+
+  React.useEffect(() => {
+    /**
+     * A unique identifier is generated for each Banner during construction. This will be used to
+     * uniquely register/unregister the banner with ancestor Banner Managers without requiring consumers to
+     * define unique identifiers themselves.
+     */
+    const uuid = uuidv4();
```
Comment (tbiethman) [inline src/banner/Banner.jsx]: Would it make sense to move this id into a ref outside the effect? Or do you want a new id with every registration?

File: src/banner/Banner.jsx [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      return () => {};
+    }
+
+    if (bannerRegistration && bannerRegistration.registerBanner) {
+      bannerRegistration.registerBanner(uuid, props);
+    }
+
+    return () => {
+      if (bannerRegistration && bannerRegistration.unregisterBanner) {
+        bannerRegistration.unregisterBanner(uuid, props.type);
```
Comment (tbiethman) [inline src/banner/Banner.jsx]: Having to unregister/re-register a banner with every prop change might lead to some flashing on slower devices if the updates aren't batched.

Have you considered only unregistering on component unmount (in a separate effect) and making registerBanner update an existing banner if one exists with the same uuid? With props like onDismiss that are typically provided with inline functions, this would happen quite often.

File: src/terra-dev-site/app/components.2/BannerPropsTable.jsx [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  }),
+  /**
+   * Nodes providing the message content for the banner. Can contain text and HTML.
+   */
+  description: PropTypes.node,
+  /**
+   * Callback function triggered when Dismiss button is clicked. The presence of this prop will cause
+   * the Dismiss button to be included on the banner.
+   */
+  onDismiss: PropTypes.func,
```
Comment (tbiethman) [inline src/terra-dev-site/app/components.2/BannerPropsTable.jsx]: I've been thinking about this class of prop with some of the work I'm doing. Our modal today has a prop called `onRequestClose` that gets called when the modal thinks it should be dismissed. The 'request' part is indicating that the modal won't _actually_ go away until the consumer actually makes it happen.

I believe the same thing is true here; until the consumer of the Banner unmounts it, it's not actually gone. My question is whether we like the 'request' nomenclature for these types of callbacks (and should include it here) or do we feel like it's too wordy and consumers will get the point without it?

File: src/banner/Banner.jsx (hunk 5 above)
Comment (emilyrohrbough) [reply src/banner/Banner.jsx]: I created this [initially with used as a ref outside of the useEffect and was getting some lint warning](https://github.com/cerner/terra-application/pull/45/commits/7e46bcac48daa0cac03e65be0bc86f07c7b31aeb#diff-dc407e6487b82f8a1ebe8f58e5088673L63). I think our usage was fine to ignore the lint warning. It seems both accomplished the same thing, unique ID when rendered. Do you think we should disable the lint rule?

EDIT. I will see how this is impacted by prop updates. I think that could cause multiple banners to be registered when it should be the same banner updating.

File: src/banner/BannerCheckpoint.jsx (hunk 3 above)
Comment (emilyrohrbough) [reply src/banner/BannerCheckpoint.jsx]: I'm fine with that. I'll update the name throughout

File: package.json (hunk 1 above)
Comment (emilyrohrbough) [reply package.json]: oh yes...

File: src/banner/Banner.jsx (hunk 2 above)
Comment (emilyrohrbough) [reply src/banner/Banner.jsx]: You are right. I will need to re-think my approach on this. Anonymous  onDismiss functions, action and customIcons are all triggering re-renders after I flattened extracted & flattened down the prop types.

File: src/application-base/ApplicationBase.jsx (hunk 4 above)
Comment (emilyrohrbough) [reply src/application-base/ApplicationBase.jsx]: I agree. I'm not sue if/when banners would be at this level when navigation & page layouts are provided in the application. We could potentially remove it from this level.

File: src/banner/Banner.jsx (hunk 5 above)
Comment (emilyrohrbough) [reply src/banner/Banner.jsx]: You are right, the prop updates triggers a re-render. Even when these are flattened, consumers may update the prop values and we would want to keep the same id. I will move back to be a ref.

File: package.json (hunk 1 above)
Comment (emilyrohrbough) [reply package.json]: fixed: https://github.com/cerner/terra-application/pull/45/commits/b70e1ed81c38d6c315d28cfd80420d69f36f28c7

File: src/banner/BannerCheckpoint.jsx (hunk 3 above)
Comment (emilyrohrbough) [reply src/banner/BannerCheckpoint.jsx]: renamed: https://github.com/cerner/terra-application/pull/45/commits/58171140c932a65d683734578709bac27efde528

File: src/banner/Banner.jsx (hunk 5 above)
Comment (emilyrohrbough) [reply src/banner/Banner.jsx]: use ref: https://github.com/cerner/terra-application/pull/45/commits/bd880f41078aeda4a25edfab35fd79421c525c3a

File: src/banner/Banner.jsx (hunk 6 above)
Comment (emilyrohrbough) [reply src/banner/Banner.jsx]: I hadn't but that is a good idea. I went ahead an moved it out into it's own useEffect hook: https://github.com/cerner/terra-application/pull/45/commits/71e60344c4d90e82148ba7e8bf976d26c80f61b4

File: src/banner/BannerProvider.jsx [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+/**
+ * The Banner Checkpoint manages prioritizing and displaying all Workflow Banners
+ * rendered within the Checkpoint Context in a list above all other content in the tree.
+ */
+const BannerProvider = ({ fitToParentIsDisabled, children }) => {
+  const registeredBanners = React.useRef({});
+  const [banners, setBanners] = React.useState([]);
+
+  const registerBanner = (bannerId, bannerProps) => {
```
Comment (tbiethman) [inline src/banner/BannerProvider.jsx]: Typically we've defined these context-related APIs inside a useMemo block to prevent use from redefining these functions (and throwing them away) with every render.

```js
const bannerProviderValue = React.useMemo(() => {
  const registerBanner = ...;
  const unregisterBanner = ...;

  return {
    registerBanner,
    unregisterBanner,
  };
}, []);
```

Comment (emilyrohrbough) [reply src/banner/BannerProvider.jsx]: updated to use memo: https://github.com/cerner/terra-application/pull/45/commits/a905caaef9296be5ffa5857c8281de69a4c9595f#diff-51cafba97ae92a564c90ef1fa3bdbfafR32

File: src/banner/Banner.jsx (hunk 2 above)
Comment (emilyrohrbough) [reply src/banner/Banner.jsx]: updated: https://github.com/cerner/terra-application/pull/45/commits/a905caaef9296be5ffa5857c8281de69a4c9595f#diff-dc407e6487b82f8a1ebe8f58e5088673R96

File: src/notification-banner/private/BannerRegistrationContext.jsx [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+/**
+ * The default value for the BannerRegistrationContext includes a flag to indicate that the value is the default value,
+ * which can be used to detect whether any matching context Providers are rendered. Also included are no-op implementations
+ * for register/unregisterBanner to minimize branching logic in the context consumers.
```
Comment (jmsv6d) [inline src/notification-banner/private/BannerRegistrationContext.jsx:7]: It doesn't look like that context has been given a default value.

**Comment by emilyrohrbough** [comment]
FYI: there are a few UX/Accessibilty decisions that are still being worked out. I am going to be adding some demos/examples for exploratory testing. No need for a review at this point of pushed changes till the blocked labels are removed.

File: src/notification-banner/private/_Banner.jsx [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+  );
+};
+
+Alert.propTypes = propTypes;
+Alert.defaultProps = defaultProps;
+Alert.Opts = {};
+Alert.Opts.Types = AlertTypes;
+
+export default Alert;
```
Comment (tbiethman) [inline src/notification-banner/private/_Banner.jsx]: Reminder to update the component name

Comment (tbiethman) [reply src/notification-banner/private/_Banner.jsx]: Doh, this is for testing isn't it?

Comment (emilyrohrbough) [reply src/notification-banner/private/_Banner.jsx]: Yeah this is added in create the test pages for the Accessibility team as the determine the expected behavior. We may or may not need to pull this forward now. Still waiting on latest UX designs as well.

File: src/notification-banner/NotificationBanner.jsx [hunk 11]
```diff
@@ -0,0 +1,84 @@
+import React from 'react';
```
Comment (emilyrohrbough) [inline src/notification-banner/NotificationBanner.jsx:1]: Ignore for now. added for Accessibility testing.

File: src/notification-banner/ProviderUsingAriaPolite.jsx [hunk 12]
```diff
@@ -0,0 +1,116 @@
+import React from 'react';
```
Comment (emilyrohrbough) [inline src/notification-banner/ProviderUsingAriaPolite.jsx]: Ignore for now. added for Accessibility testing.

File: src/notification-banner/private/Banner.module.scss [hunk 13]
```diff
@@ -0,0 +1,163 @@
+:local {
```
Comment (emilyrohrbough) [inline src/notification-banner/private/Banner.module.scss]: Ignore for now. added for Accessibility testing.

File: src/notification-banner/private/BannerRegistrationContext.jsx (hunk 9 above)
Comment (emilyrohrbough) [reply src/notification-banner/private/BannerRegistrationContext.jsx:7]: Would you expect a default value of `React.createContext({});`?

File: src/notification-banner/private/_Banner.jsx [hunk 14]
```diff
@@ -0,0 +1,205 @@
+import React, { useState } from 'react';
```
Comment (emilyrohrbough) [inline src/notification-banner/private/_Banner.jsx]: Ignore for now. added for Accessibility testing.

File: src/notification-banner/useNotificationBanners.jsx [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+const useNotificationBanners = () => {
+  /** The registeredBanners ref maintains collection of banner ids and banner props registered to the  NotificationBannerProvider. */
+  const registeredBanners = React.useRef({});
+
+  /**
+   * The updateBannerState ref stores the update state function used to manage the banner rendered in the NotificationBanners component.
+   * This ties the state updates to the `useNotificationBanners` hook, while allowing the NotificationBanners to be rendered above or below
+   * the NotificationBannerProvider.
+   */
+  const updateBannerState = React.useRef();
```
Comment (emilyrohrbough) [inline src/notification-banner/useNotificationBanners.jsx]: @tbiethman I went forward with this approach. It seems to work great and have yet to notice any unexpected consequences. Add a bunch of docs to hopefully explain what is going on. Will see what other think.

File: src/terra-dev-site/app/components.2/NotificationBannerProvider.app.mdx [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+- Error
+- Warning
+- Unsatisfied
+- Unverified
+- Advisory
+
+## Usage
+
+```jsx
+import { useNotificationBanners } from 'terra-application/lib/notification-banner';
```
Comment (emilyrohrbough) [inline src/terra-dev-site/app/components.2/NotificationBannerProvider.app.mdx]: reminder to myself. Update this doc if approach is approved by others.

File: tests/jest/disclosure-manager/DisclosureContainer.test.jsx [hunk 17]
```diff
@@ -17,17 +17,4 @@ describe('DisclosureContainer', () => {
     const wrapper = shallow(disclosureContainer);
     expect(wrapper).toMatchSnapshot();
   });
-
-  it('should render the provided children', () => {
```
Comment (emilyrohrbough) [inline tests/jest/disclosure-manager/DisclosureContainer.test.jsx]: duplicate test: https://github.com/cerner/terra-application/pull/45/files/a87c9663e3531db82e6411e893e22f4e41678f3d?file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.jsx&file-filters%5B%5D=.mdx&file-filters%5B%5D=.png&file-filters%5B%5D=.scss#diff-5c8f307acd5212386bcde04e991a12d5R8

**Comment by scottwilmarth** [comment]
Emily Ogle, Rebecca Deitch, and I have reviewed the latest banner notification in NVDA, JAWS, and VoiceOver and all seems to be working and on the right path. We give it thumbs up.

File: src/notification-banner/NotificationBanner.jsx [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+
+  React.useEffect(() => {
+    /**
+     * If the bannerRegistration value is the ProviderRegistrationContext's default value,
+     * then there is not a matching BannerProvider above it in the hierarchy.
+     * This is possible but likely not intentional, so the component warns.
+     */
+    if (!bannerRegistration && process.env.NODE_ENV !== 'production') {
+      // eslint-disable-next-line no-console
+      console.warn('A NotificationBanner was not rendered within the context of a NotificationBannerProvider. If this is unexpected, validate that the expected version of the terra-application package is installed.');
```
Comment (tbiethman) [inline src/notification-banner/NotificationBanner.jsx]: When we solidify the different framework layers in v2, I think we should start throwing errors here rather than warn. If a component renders a NotificationBanner in the wrong spot, and we cannot fulfill the promise that a banner will be presented for them, we should make the problem obvious. We could make a `useValidatedContext` custom hook or something that will throw if the context isn't defined and passthrough otherwise.

I know we warn a few other places today for these context misses, though, so I'll leave updating this up to you.

File: src/notification-banner/private/BannerRegistrationContext.jsx (hunk 9 above)
Comment (tbiethman) [reply src/notification-banner/private/BannerRegistrationContext.jsx:7]: I'd prefer to have an undefined value so that we can detect (and throw an error) when there's no provider in place.

File: src/terra-dev-site/app/components.2/BannerPropsTable.jsx (hunk 7 above)
Comment (emilyrohrbough) [reply src/terra-dev-site/app/components.2/BannerPropsTable.jsx]: updated to onRequestClose: https://github.com/cerner/terra-application/pull/45/commits/c36f75662652b37024088bf93d033b73d9d4c71b

**Comment by jeremyfuksa** [comment]
Everything is looking alright from a general UX perspective. Is this ready for a review stamp (aside from the Travis build)?

**Comment by emilyrohrbough** [comment]
@jeremyfuksa according to @neilpfeiffer the designs are being held up to review API consistencies with the MVB of notification-dialog: https://github.com/cerner/terra-framework/pull/1105

**Comment by jeremyfuksa** [comment]
> @jeremyfuksa according to @neilpfeiffer the designs are being held up to review API consistencies with the MVB of notification-dialog: [cerner/terra-framework#1105](https://github.com/cerner/terra-framework/pull/1105)

Holding tight, then.

**Comment by emilyrohrbough** [comment]
@cerner/terra this is ready for review.

File: src/notification-banner/NotificationBanner.jsx (hunk 18 above)
Comment (emilyrohrbough) [reply src/notification-banner/NotificationBanner.jsx]: I have updated this to throw errors. 👍

**Comment by PaulSolDev** [comment]
Functional review complete. No issues identified.

File: src/notification-banner/NotificationBanner.jsx [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+    /**
+     * If the bannerRegistration value is the ProviderRegistrationContext's default value,
+     * then there is not a matching BannerProvider above it in the hierarchy.
+     * This is possible but likely not intentional, so the component warns.
+     */
+    if (!bannerRegistration) {
+      throw new Error('A NotificationBanner was not rendered within the context of a NotificationBannerProvider. If this is unexpected, validate that the expected version of the terra-application package is installed.');
+    }
+
+    if (bannerRegistration && bannerRegistration.registerNotificationBanner) {
```
Comment (tbiethman) [inline src/notification-banner/NotificationBanner.jsx]: Given your check on line 76, I don't think you need the presence checks here or below for the unregister. We'd consider any of these missing a critical error scenario.

File: src/notification-banner/NotificationBanner.jsx [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+}) => {
+  /**
+   * A unique identifier is generated for each Banner during construction. This will be used to
+   * uniquely register/unregister the banner with ancestor Banner Managers without requiring consumers to
+   * define unique identifiers themselves.
+   */
+  const bannerRegistration = React.useContext(BannerRegistrationContext);
+
+  React.useEffect(() => {
+    const uuid = uuidv4();
```
Comment (tbiethman) [inline src/notification-banner/NotificationBanner.jsx]: Did generating the uuid per-registration (vs per instance, stored in a ref) make registration logic simpler? I like how the status provider had two effects, one to handle registration (and re-registration, if values changed) and another that handles unregistration that will only execute once: https://github.com/cerner/terra-application/blob/main/src/application-status-overlay/ApplicationStatusOverlay.jsx#L41-L51

Having to unregister just to register again seems heavy, even if it doesn't have a visual impact. It's also nice to be able to tie an instance of this component to a single identifier for debugging purposes. 🤷‍♂️ Thoughts?

File: src/notification-banner/private/useNotificationBanners.jsx [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+    children: PropTypes.node,
+  };
+
+  /**
+   * Renders a list of prioritized notification banners.
+   */
+  const NotificationBanners = () => {
+    const theme = React.useContext(ThemeContext);
+    const [banners, setBanners] = React.useState([]);
+    updateBannerState.current = setBanners;
```
Comment (tbiethman) [inline src/notification-banner/private/useNotificationBanners.jsx]: 👀  we might want to put some comments here about what this line is doing

File: src/terra-dev-site/app/Application.1.app.mdx [hunk 22]
```diff
@@ -65,6 +65,12 @@ Presenting status views to show the current state of the page is a common applic
 - [Application Status View API](/application/terra-application/components/application-status-overlay)
 - [Application Status View Provider API](/application/terra-application/components/application-status-overlay-provider)
 
+### Notification Banners
+
+Notification Banners are used to bring the user's attention to important displayed information. This common interface pattern requires strategic placement and ordering on the application screen to be effective. Orion Application provides banner management to align with these needs.
```
Comment (tbiethman) [inline src/terra-dev-site/app/Application.1.app.mdx]:
```suggestion
Notification Banners are used to bring the user's attention to important displayed information. This common interface pattern requires strategic placement and ordering on the application screen to be effective. Terra Application provides banner management to align with these needs.
```

File: src/notification-banner/private/useNotificationBanners.jsx [hunk 23]
```diff
@@ -0,0 +1,190 @@
...
+          );
+        })}
+      </div>
+    );
+  };
+
+  const useNotificationBannerExports = React.useRef({
+    NotificationBannerProvider,
+    NotificationBanners,
+  });
```
Comment (tbiethman) [inline src/notification-banner/private/useNotificationBanners.jsx]: With this implementation, we're generating new NotificationBannerProvider and NotificationBanners with every update, but only ever returning the first ones that get generated. It's a minor waste of computation time, but you could wrap this creation in a useMemo block to limit this.

**Review by jeremyfuksa — APPROVED** [review]
Everything appears to be working as expected from a UX standpoint.

File: src/notification-banner/NotificationBanner.jsx (hunk 20 above)
Comment (emilyrohrbough) [reply src/notification-banner/NotificationBanner.jsx]: > Did generating the uuid per-registration (vs per instance, stored in a ref) make registration logic simpler?  

it didn't affect the registration logic, the linter was complaining and I didn't notice any adverse affect when I moved it. I see your point on only un-registering once. I had logic similar to that at one point and then changed it. I'll work on updating this to be lighter.

File: src/notification-banner/NotificationBanner.jsx (hunk 19 above)
Comment (emilyrohrbough) [reply src/notification-banner/NotificationBanner.jsx]: fixed: https://github.com/cerner/terra-application/pull/45/commits/b0d9709f34d39188fdfb84b3015ee5f6a9ae00a3#diff-1c1abe24b9b3aaa6ad0c051069883f78L80

File: src/notification-banner/NotificationBanner.jsx (hunk 20 above)
Comment (emilyrohrbough) [reply src/notification-banner/NotificationBanner.jsx]: fixed: https://github.com/cerner/terra-application/pull/45/commits/de735223046eaabc526799ffde4a1caedc8db758#diff-1c1abe24b9b3aaa6ad0c051069883f78R67

File: src/notification-banner/private/useNotificationBanners.jsx (hunk 21 above)
Comment (emilyrohrbough) [reply src/notification-banner/private/useNotificationBanners.jsx]: added: https://github.com/cerner/terra-application/pull/45/commits/1efd4b0fe6254e30f83bc4e54571fd2137a9ae27#diff-7732e149c9945ef93b0d5632b181cef4R111

File: src/terra-dev-site/app/Application.1.app.mdx (hunk 22 above)
Comment (emilyrohrbough) [reply src/terra-dev-site/app/Application.1.app.mdx]: fixed: https://github.com/cerner/terra-application/pull/45/commits/de735223046eaabc526799ffde4a1caedc8db758#diff-a1483c6896b3e882702f7ab6a0348b91R70

File: src/notification-banner/private/useNotificationBanners.jsx [hunk 24]
```diff
@@ -0,0 +1,182 @@
...
+        updateBannerState.current({ ...registeredBanners.current });
+      }
+    };
+
+    return {
+      /**
+       * Provides the Banner Registration Context to its children.
+       */
+      NotificationBannerProvider: ({ children }) => ( // eslint-disable-line react/prop-types
+        <BannerRegistrationContext.Provider value={{ registerNotificationBanner, unregisterNotificationBanner }}>
```
Comment (tbiethman) [inline src/notification-banner/private/useNotificationBanners.jsx:90]: I believe this will still create a new value object with every update. You could move the value into a const outside the component (like to line 84) to avoid this.

Comment (emilyrohrbough) [reply src/notification-banner/private/useNotificationBanners.jsx:90]: This hook is only ran on mount so I'm not sure it matters if this is a declared as a const or created in-line. I can update this though.

Comment (tbiethman) [reply src/notification-banner/private/useNotificationBanners.jsx:90]: The NotificationBannerProvider component is created once, but the NotificationBannerProvider is executed every time it gets rendered, and as written each execution will create a new value object. Moving the object to the outer scope ensures a consistent reference is used with each render.

**Review by tbiethman — APPROVED** [review]

File: src/notification-banner/index.js [hunk 25]
```diff
@@ -0,0 +1,3 @@
+import NotificationBanner from './NotificationBanner';
+
+export default NotificationBanner;
```
Comment (mjhenkes) [inline src/notification-banner/index.js]: This doesn't matter, but you could shorten this line
```suggestion
import { default as NotificationBanner } from './NotificationBanner';

```

Comment (mjhenkes) [reply src/notification-banner/index.js]: hold on i got this backwards

Comment (mjhenkes) [reply src/notification-banner/index.js]: there might be a way to one line this but i've already spent too much time on it 😁

File: src/notification-banner/private/useNotificationBanners.jsx [hunk 26]
```diff
@@ -0,0 +1,182 @@
...
+
+const cx = classNamesBind.bind(styles);
+
+/**
+ * The `useNotificationBanners` Hook manages registering and prioritizing Notification Banners
+ * rendered within the Notification Banner Context.
+ *
+ * Returns:
+ *   - NotificationBannerProvider - React Context Provider - Provides the Banner Registration Context to its children.
+ *         This allows any NotificationBanner registered beneath it to be displayed in teh NotificationBanner's list.
```
Comment (mjhenkes) [inline src/notification-banner/private/useNotificationBanners.jsx]:
```suggestion
 *         This allows any NotificationBanner registered beneath it to be displayed in the NotificationBanner's list.
```

File: src/notification-banner/private/organizeBannersByPriority.js [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * @returns List of prioritized banners.
+ */
+const organizeBannersByPriority = (banners, themeName) => {
+  if (!banners) {
+    return [];
+  }
+
+  let priorityOrder = defaultBannerPriority;
+  if (themeName === 'orion-fusion-theme') {
```
Comment (ryanthemanuel) [inline src/notification-banner/private/organizeBannersByPriority.js:34]: Exciting!!!

File: src/notification-banner/private/organizeBannersByPriority.js [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+    priorityOrder = fusionBannerPriority;
+  }
+
+  const bannersList = priorityOrder.reduce((coll, variant) => {
+    if (banners[variant]) {
+      // eslint-disable-next-line compat/compat
```
Comment (ryanthemanuel) [inline src/notification-banner/private/organizeBannersByPriority.js]: Why do we need this disable?

File: src/terra-dev-site/app/components.2/NotificationBanner.app.mdx [hunk 29]
```diff
@@ -0,0 +1,74 @@
...
+
+# Notification Banner
+
+A Notification Banner should be rendered by a component that needs to render a notification with information that needs to be brought to the user's attention. This is a non-disruptive notification that does not require immediate input or attention by the user.
+
+When rendered, the Notification Banner registers itself with the framework so it prioritized correctly when any other rendered notification banners before it is displayed in a list above all other content.
+
+## Banner Variants
+
+The Notification Banner can be used to bring awarness to the a user’s risk and non-risk situations that may occur within an application; there are multiple variants provided for common situtations.
```
Comment (ryanthemanuel) [inline src/terra-dev-site/app/components.2/NotificationBanner.app.mdx]:
```suggestion
The Notification Banner can be used to bring awareness to a user’s risk and non-risk situations that may occur within an application; there are multiple variants provided for common situations. 
```

File: src/terra-dev-site/app/components.2/NotificationBanner.app.mdx [hunk 30]
```diff
@@ -0,0 +1,74 @@
+import PropsTable from './NotificationBannerPropsTable?dev-site-props-table';
+import NotificationBannerExample from './example/NotificationBannerExample?dev-site-example';
+
+# Notification Banner
+
+A Notification Banner should be rendered by a component that needs to render a notification with information that needs to be brought to the user's attention. This is a non-disruptive notification that does not require immediate input or attention by the user.
+
+When rendered, the Notification Banner registers itself with the framework so it prioritized correctly when any other rendered notification banners before it is displayed in a list above all other content.
```
Comment (ryanthemanuel) [inline src/terra-dev-site/app/components.2/NotificationBanner.app.mdx]:
```suggestion
When rendered, the Notification Banner registers itself with the framework so it is prioritized correctly when any other rendered notification banners before it are displayed in a list above all other content.
```

File: src/terra-dev-site/app/components.2/NotificationBanner.app.mdx [hunk 31]
```diff
@@ -0,0 +1,74 @@
...
+
+## Banner Variants
+
+The Notification Banner can be used to bring awarness to the a user’s risk and non-risk situations that may occur within an application; there are multiple variants provided for common situtations. 
+
+- Notification Banners show for risk situations are intended to notify the user of hazards, such as risks to patient safety or causing financial harm, and provide options to mitigate or remove the hazard. There are three variants of risk oriented notification banners, based on the severity: `hazard-high`, `hazard-medium` and `hazard-low`. These variants are rigid to provide consistent icon and signal word pairs to prevent accidental confusion by the user of the banner's severity.
```
Comment (ryanthemanuel) [inline src/terra-dev-site/app/components.2/NotificationBanner.app.mdx]:
```suggestion
- Notification Banners shown for risk situations are intended to notify the user of hazards, such as risks to patient safety or causing financial harm, and provide options to mitigate or remove the hazard. There are three variants of risk oriented notification banners, based on the severity: `hazard-high`, `hazard-medium` and `hazard-low`. These variants are rigid to provide consistent icon and signal word pairs to prevent accidental confusion by the user of the banner's severity.
```

File: src/terra-dev-site/app/components.2/NotificationBanner.app.mdx [hunk 32]
```diff
@@ -0,0 +1,74 @@
...
+- The custom variant allows for less common standardized non-hazard situations, requiring an icon and signal word be provided in addition to the common props. By choosing to create a custom Notification Banner, you are responsible for providing an icon that is themed correctly for the terra-default-theme, orion-fusion-theme, and clinical-lowlight-theme. See Terra’s [Theme Strategy Guide](/guides/terra-application/theme-strategy) for more info.
+
+## Banner Priority Order
+
+Notification Banners are prioritized by variant and in the following order:
+
+1. `hazard-high` - displays a critical notification for current emergencies or situations of high severity that need to be addressed promptly
+2. `hazard-medium` - recommended to show moderately urgent notifications that are cautionary in nature and forecast potential risk
+3. `hazard-low` - recommended to provide the user with guidance and/or advice for completing an action
+4. `error` - recommended to provide feedback of a negative system status such as failures, inabilities, or limitations
```
Comment (ryanthemanuel) [inline src/terra-dev-site/app/components.2/NotificationBanner.app.mdx]: Possibly:

```suggestion
4. `error` - recommended to provide feedback of a negative system status such as failures or limitations of the system
```

File: src/terra-dev-site/app/components.2/NotificationBanner.app.mdx [hunk 33]
```diff
@@ -0,0 +1,74 @@
...
+<PropsTable />
+
+## Example
+
+<NotificationBannerExample />
+
+## Testing
+
+The Notification Banner has the following data attribute available for browser tests:
+- The data attribute, `data-terra-application-notification-banner` can be used to validate the expect banner exists
```
Comment (ryanthemanuel) [inline src/terra-dev-site/app/components.2/NotificationBanner.app.mdx]:
```suggestion
- The data attribute, `data-terra-application-notification-banner` can be used to validate the expected banner exists
```

File: src/terra-dev-site/test/notification-banner/ExampleForAccessibilityTesting.jsx [hunk 34]
```diff
@@ -0,0 +1,113 @@
...
+        />
+      )}
+      {showHazardLowBanner && <NotificationBanner variant="hazard-low" id={`hazard-low-banner-${id}`} />}
+      <p>The error banner will appear after 10 seconds.</p>
+      <button
+        onClick={() => setShowErrorBanner(!showErrorBanner)}
+        type="button"
+        id={`toggle-error-banner-${id}`}
+      >
+        Show/Hide Error Banner
```
Comment (ryanthemanuel) [inline src/terra-dev-site/test/notification-banner/ExampleForAccessibilityTesting.jsx:62]: Could we flex the text of these buttons based on whether or not things are already hidden shown?  Could make the examples a little easier to interact with.

File: src/terra-dev-site/app/components.2/NotificationBanner.app.mdx [hunk 35]
```diff
@@ -0,0 +1,74 @@
+import PropsTable from './NotificationBannerPropsTable?dev-site-props-table';
```
Comment (mjhenkes) [inline src/terra-dev-site/app/components.2/NotificationBanner.app.mdx]: Why can't you import the props from the notification banner component? Because it returns null?

File: src/terra-dev-site/app/utilities.2/Logger.app.mdx [hunk 36]
```diff
@@ -32,7 +32,7 @@ initializeLogger({
 
 ### Logger
 
-|method|syntax|Default Behaviour|
+|method|syntax|Default Behavior|
```
Comment (mjhenkes) [inline src/terra-dev-site/app/utilities.2/Logger.app.mdx:35]: British spelling > American

File: src/terra-dev-site/app/components.2/NotificationBanner.app.mdx (hunk 35 above)
Comment (emilyrohrbough) [reply src/terra-dev-site/app/components.2/NotificationBanner.app.mdx]: yes because of null

Comment (mjhenkes) [reply src/terra-dev-site/app/components.2/NotificationBanner.app.mdx]: react doc gen is an imperfect tool for this use case 😕

**Review by mjhenkes — APPROVED** [review]

File: src/terra-dev-site/test/notification-banner/ExampleForAccessibilityTesting.jsx (hunk 34 above)
Comment (emilyrohrbough) [reply src/terra-dev-site/test/notification-banner/ExampleForAccessibilityTesting.jsx:62]: At this time, I'd prefer to make this change in a follow up PR.

**Review by ryanthemanuel — APPROVED** [review]
